Arduino Uno R3/examples/CN0411_example/AD7124.cpp

```diff
@@ -0,0 +1,498 @@
+*
+* @param device - The handler of the instance of the driver.
+* @param reg - Register structure holding info about the register to be written
+*
+* @return Returns 0 for success or negative error code.
+*******************************************************************************/
+int32_t AD7124_NoCheckWriteRegister(struct ad7124_device *device,
+				    struct ad7124_st_reg reg)
+{
+	int32_t ret      = 0;
+	int32_t regValue = 0;
+	uint8_t wrBuf[8] = {0, 0, 0, 0, 0, 0, 0, 0};
+	uint8_t i        = 0;
+	uint8_t crc8     = 0;
+
+	if(!device)
+		return INVALID_VAL;
+
+	/* Build the Command word */
+	wrBuf[0] = AD7124_COMM_REG_WEN | AD7124_COMM_REG_WR |
+		   AD7124_COMM_REG_RA(reg.addr);
+
+	/* Fill the write buffer */
+	regValue = reg.value;
+	for(i = 0; i < reg.size; i++) {
+		wrBuf[reg.size - i] = regValue & 0xFF;
+		regValue >>= 8;
+	}
+
+	/* Compute the CRC */
+	if(device->useCRC != AD7124_DISABLE_CRC) {
+		crc8 = AD7124_ComputeCRC8(wrBuf, reg.size + 1);
+		wrBuf[reg.size + 1] = crc8;
+	}
+
+	/* Write data to the device */
+	ret = SPI_Write(device->slave_select_id,
+			wrBuf,
+			(device->useCRC != AD7124_DISABLE_CRC) ? reg.size + 2
+			: reg.size + 1);
+
+	return ret;
+}
+
+/***************************************************************************//**
+* @brief Reads the value of the specified register only when the device is ready
+*        to accept user requests. If the device ready flag is deactivated the
+*        read operation will be executed without checking the device state.
+*
+* @param device - The handler of the instance of the driver.
+* @param pReg - Pointer to the register structure holding info about the
+*               register to be read. The read value is stored inside the
+*               register structure.
+*
+* @return Returns 0 for success or negative error code.
+*******************************************************************************/
+int32_t AD7124_ReadRegister(struct ad7124_device *device,
+			    struct ad7124_st_reg* pReg)
+{
+	int32_t ret;
+
+	if (pReg->addr != AD7124_ERR_REG && device->check_ready) {
+		ret = AD7124_WaitForSpiReady(device, device->spi_rdy_poll_cnt);
+		if (ret < 0)
+			return ret;
+	}
+	ret = AD7124_NoCheckReadRegister(device, pReg);
+
+	return ret;
+}
+
+/***************************************************************************//**
+* @brief Writes the value of the specified register only when the device is
+*        ready to accept user requests. If the device ready flag is deactivated
+*        the write operation will be executed without checking the device state.
+*
+* @param device - The handler of the instance of the driver.
+* @param reg - Register structure holding info about the register to be written
+*
+* @return Returns 0 for success or negative error code.
+*******************************************************************************/
+int32_t AD7124_WriteRegister(struct ad7124_device *device,
+			     struct ad7124_st_reg pReg)
```

Comment:
this seems to be a re-occuring issue with this driver, where data gets passed by value vs reference ;
[ we may need to fix the no-OS driver i guess ]
ideally it should be reference, because otherwise it copies `ad7124_st_reg` on every call, which is a bit much